Projects/Modules/Documentarian.MicrosoftDocs/Source/Public/Functions/Get-CmdletXref.ps1

```diff
@@ -1,52 +1,69 @@
-﻿# Copyright (c) Microsoft Corporation.
+# Copyright (c) Microsoft Corporation.
 # Licensed under the MIT License.
 
+using namespace System.Management.Automation
+
 function Get-CmdletXref {
     <#
     .SYNOPSIS
     Gets a cross-reference link for a command.
     #>
     [CmdletBinding()]
-    [OutputType('System.String[]')]
+    [OutputType([string[]])]
     param(
-        # The name of the command to get a cross-reference link for.
+        # The name of the Command or a CommandInfo instance to get a cross-reference link for.
         [Parameter(Mandatory, ValueFromPipeline)]
-        [string[]]$Name
+        [object[]] $Command
```

Comment:
Having a think about this, I would prefer it if we updated the implementation to have two mutually exclusive parameter sets - one for string names, one for **CommandInfo** objects. Both should be passable through the pipeline. That will ensure we get both input types in the reference documentation without having to write a longer description clarifying the behavior. It'll also help us to avoid needing to handle when a user submits other types, like integers or hashtables.

The alternative would be to define a new argument transformer.

Reply:
@michaeltlombardi Yes, let's go with 2 parameter sets: 'ByName', 'ByCommand'. The docs need to be updated for the new parameter sets and new examples for each.

Reply:
Both options make sense, 2 mutually exclusive parameters `[string]` and `[CommandInfo]` or a single parameter `[CommandInfo]` decorated with `ArgumentTransformationAttribute`. Let me know which one you prefer @michaeltlombardi & @sdwheeler. If going with the latter option, I assume the class definition would go into the `Classes` folder right? 😄 just making sure.

Reply:
It would! I think for now let's go with the parameter sets, because that _should_ be less work for you, I believe. 😅

Reply:
I'm fine either way, the implementation for `ArgumentTranformationAttribute` would be more or less like this:

```powershell
using namespace System.Management.Automation

class CommandInfoArgTransformation : ArgumentTransformationAttribute {
    [Object] Transform([EngineIntrinsics] $engineIntrinsics, [Object] $inputData) {
        $cmd = $inputData

        if ($inputData -isnot [CommandInfo]) {
            $cmd = $engineIntrinsics.InvokeCommand.GetCommand(
                $inputData,
                [CommandTypes]::All)
        }

        if (-not $cmd) {
            throw [CommandNotFoundException]::new("Unable to find command '$inputData'.")
        }

        # using arg transform would also allow to move the logic for alias here, i.e.:
        if ($cmd.CommandType -eq [CommandTypes]::Alias) {
            return $cmd.ResolvedCommand
        }

        return $cmd
    }
}

function Test-ArgTransformation {
    param(
        [CommandInfoArgTransformation()]
        [CommandInfo] $Command
    )

    $Command
}

# caveat: makes a terminating error here:
Test-ArgTransformation doesnotexist
Test-ArgTransformation gci
Test-ArgTransformation Get-ChildItem
Test-ArgTransformation (Get-Command Get-ChildItem)
```

Reply:
🤔 Okay, if you're up for it, I'm on board for the argument transformer. We might end up having other commands where we want to operate on cmdlets as either strings or commandinfo objects.

Reply:
Hi @michaeltlombardi and @sdwheeler the Transformation Class is implemented. I tried to make it in smaller commits so (hopefully) it's easier to review.

As aside, if you would like, it would be very simple to use [`CompletionCompleters.CompleteCommand` method](https://learn.microsoft.com/zh-tw/dotnet/api/system.management.automation.completioncompleters.completecommand?view=powershellsdk-1.1.0#system-management-automation-completioncompleters-completecommand(system-string)) in case you want to add completions to this function. Here I have a sample of the class definition: https://github.com/santisq/PSParallelPipeline/blob/main/Module/private/CommandCompleter.ps1